Code snipppets D-F v2

mm-broadcom opened the pull request:
Second PR for the D-F snippets as the previous one (#2794) was getting long and there were a few issues brought up that were not addressed prior to getting merged in.  

## How Has This Been Tested?

Ran and checked within the extension running locally.

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

mm-broadcom committed:
Start of work on snippets

mm-broadcom committed:
More snippets

mm-broadcom committed:
Add APPLY descriptive text

mm-broadcom committed:
Improve snippets

mm-broadcom committed:
More terms and definitions

mm-broadcom committed:
Move, revise and add more snippets

mm-broadcom committed:
"D" Snippets

mm-broadcom committed:
E and F snippets

mm-broadcom committed:
Update snippet test counts

mm-broadcom committed:
Fix snippet test check

mm-broadcom committed:
Update integration test

mm-broadcom committed:
Ran prettier

mm-broadcom committed:
More test fixes

mm-broadcom committed:
Adjust integration test

mm-broadcom committed:
Another integration test fix

mm-broadcom committed:
Adjust snippet counts for tests.

hohy committed:
fix: Use default dialect name for SQL dialect (#2770)

mm-broadcom committed:
Squashed commit of the following:

commit 4e7236ef77b69ec91ad7e4a489bf2a4cc0532d36
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Tue Mar 11 10:10:29 2025 -0400

    Add missed requested change.

commit 030f5fc50568a8cad4069947edf31c8f5f66f1cf
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Tue Mar 11 10:05:38 2025 -0400

    Implemented requested changes

commit bf39023ac1d7e16d93e230e3cc17bcddea9c7ed8
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Mon Mar 10 12:06:55 2025 -0400

    Make options ALL-CAPS

commit 0bf12860c456f8dcf126b1ebadbf3c4c044431ba
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Fri Mar 7 07:47:17 2025 -0500

    Update BIT-TO-CHAR description

commit 1e08e0e427d5461f00e4d0fcaa4927cd485d6bb0
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 12:34:41 2025 -0500

    Fix snippet count tests

commit 014c287e139ba23e6beaf4306f1c62fff733273d
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 12:24:38 2025 -0500

    Remove ANYCASE snippet

commit 48dc403133764be4748fbb34588ada03404c65cd
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 12:08:11 2025 -0500

    Remove ALPHANUMERIC snippet

    Removed due to it not existing as a special class in IBM-based COBOL.

commit f7c37d00168e623ffca9fcea7184281f5342ec69
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 11:30:36 2025 -0500

    Ran prettier

commit ac29ca8d2dbc2870128a469c7694b3f011ea6205
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 11:20:43 2025 -0500

    Adjust failing test

commit c984fb8a12775eb7250056e3654af0f83f31ef3d
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 11:00:59 2025 -0500

    Fix BIT-TO-CHAR description.

commit ccf7967c071e794216cf08aea5dc7b5d643e922a
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:53:40 2025 -0500

    Added REPLACING keyword to COPY

commit b02ac6a3a8237dcf7d38f7aa51bd3c02325e40b7
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:38:51 2025 -0500

    Improved ASCENDING

commit d52e329d6ae9481051aea10d11e3220342ffb835
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:16:05 2025 -0500

    Ran prettier

commit b6e3caaf5f96ea4a69c6f44244d48b31354435fc
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:15:08 2025 -0500

    Improve ADVANCING

commit 38d2bdb6833ed8b6fce7e627a71519da77b10c71
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:08:09 2025 -0500

    Ran prettier

commit 93689d37ec591af1aa430081af61fde3195258a6
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:06:40 2025 -0500

    Fixed ACCESS MODE's body.

commit bfae2af2c4d449bdfeaabb1d208b1f715b053259
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:04:24 2025 -0500

    Change Overflow to SIZE ERROR

commit 5a452d20f434eaa5e394848e09797de330500899
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 10:00:45 2025 -0500

    Made ACCESS MODE intent clearer

commit 64591deba508efe04c8772c432cce009d6d316ba
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Thu Mar 6 09:47:01 2025 -0500

    Improve descriptions, switch forced spaces to \t

commit 5028f16039db873f5d06fd010c17404f1c8bada3
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Tue Mar 4 08:45:15 2025 -0500

    Remove backticks

    Remove backticks. They were added due to the assumption that the description used markdown formatting. It would have resulted in those portions of text looking like code, making it easier to read/follow.

commit c8ca916f2e0054dc0f58efa7582a3f5876401916
Author: Michael Messer <michael.messer@broadcom.com>
Date:   Fri Feb 28 14:56:27 2025 -0500

    Improve formatting of the descriptions.

mm-broadcom committed:
Adjust tests after merge from A-C branch

mm-broadcom committed:
Remove backticks

mm-broadcom committed:
Replaced a non-descriptive description

zeibura committed:
Update README.md (#2743)

zeibura committed:
Update README.md (#2701)

Nurkambay committed:
Move CCF analysis to COBOL LS (#2755)

* Move CCF analysis to COBOL LS

hohy committed:
fix: Prevent zowe profile popup for local only programs (#2771)

amanPras committed:
cache custom COBOL LS configuration (like processor groups etc) (#2766)

* fix: do not read processor group configuration for virtual files

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* test: adjust tests

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* fix: experimental cache setup for the cobol LS configuration

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* refactor: rename maps

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* fix: fix cache bug and adjust tests

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* revert: revert un necessary changes

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* refactor: refactor to pass the build

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* refactor: small code refactor

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* fix: make code changes as suggested in PR review

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* fix: update processor config as suggested in PR

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* chore: fix styling

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* fix: fix PR review comments

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* fix: update cache implementation for Bridge4Git

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* fix: remove unnecessary code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

* Update clients/cobol-lsp-vscode-extension/src/services/BridgeForGitLoader.ts

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

* Update clients/cobol-lsp-vscode-extension/src/services/BridgeForGitLoader.ts

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

* chore: rebase with development

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

---------

Signed-off-by: amanprashant <aman.prashant@broadcom.com>
Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

ishche committed:
feat: support IN/OF for procedures in TestCaseEngine grammar

ishche committed:
chore: prepare procedure model in usecase engine

ishche committed:
remove unused import

ishche committed:
chore: inline lambda call

ilidio-lopes committed:
tests/automation: XML generate automation (#2716)

* xml generate automation

* pr comments

ishche committed:
chore: update use case engine grammar to make it possible to track procedure declarations

ishche committed:
fix: usecase grammar issues

ishche committed:
fix: procedure division and declaratives ens in multitokens for usecase engine

Nurkambay committed:
Fix tests (#2783)

Nurkambay committed:
Create new CCF API (#2779)

* Add new CCF API

ishche committed:
fix: bring missing tests back to build

ishche committed:
chore: UseCase engine tracks sections for paragraph definitions

hohy committed:
test: make copybooks completions test more stable (#2786)

ishche committed:
chore: add section/paragraph definition use case for the testing framework

slavek-kucera committed:
fix: Avoid dangling rejected promises in control flow analysis (#2788)

amanPras committed:
chore: upgrade junit from 5.11.0-M2 to 5.11.2

A regression was introduced in 5.11.0-M2 as part of
PR https://github.com/junit-team/junit5/commit/bc87b0762717f35c49169bc97d61b338cb382c21

This resulted in failures to create usecase test files when run with "usecase.test.repo.dir" flag (https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/pull/2486).

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

KutluOzel-b committed:
feat: Enable remote copybook definitions in processor groups (#2758)

mm-broadcom committed:
Merge branch 'development' into Code-Snipppets_D-F

mm-broadcom committed:
Ran Prettier

mm-broadcom committed:
Merge branch 'Code-Snipppets_D-F' of https://github.com/mm-broadcom/che-che4z-lsp-for-cobol into Code-Snipppets_D-F

mm-broadcom committed:
Adjust integration test

mm-broadcom committed:
Adjust integration test

mm-broadcom committed:
Adjust test

mm-broadcom committed:
Adjust test

mm-broadcom committed:
Adjust test

mm-broadcom committed:
Adjust test

mm-broadcom committed:
Change links to EN and from 6.2 to 6.4

mm-broadcom committed:
Change links to EN and from 6.2 to 6.4

mm-broadcom committed:
Implemented various fixes

mm-broadcom committed:
Implemented various fixes

mm-broadcom committed:
Change paragraph to section in the declaratives snippet

mm-broadcom committed:
Change paragraph to section in the declaratives snippet

mm-broadcom committed:
Fix snippet count tests

mm-broadcom committed:
Fix snippet count tests

mm-broadcom committed:
Move section/paragraph requirements to start of description

mm-broadcom committed:
Move section/paragraph requirements to start of description

mm-broadcom committed:
Ran prettier

mm-broadcom committed:
Ran prettier

mm-broadcom committed:
Minor fixes

Fixed some descriptions, misspellings and other oversights.

mm-broadcom committed:
Minor fixes

Fixed some descriptions, misspellings and other oversights.

mm-broadcom committed:
Add snippet for COMPUTE, fix grammar

mm-broadcom committed:
Add snippet for COMPUTE, fix grammar

mm-broadcom committed:
Adjust snippet count tests

mm-broadcom committed:
Adjust snippet count again

mm-broadcom committed:
Improve descriptions

mm-broadcom committed:
Undo rebase

mm-broadcom committed:
Merge branch 'Code-Snipppets_D-F' of https://github.com/mm-broadcom/che-che4z-lsp-for-cobol into Code-Snipppets_D-F

mm-broadcom committed:
Made DATE-FIELD a tab-stop for AT DATE

mm-broadcom committed:
Fix errors and omissions

mm-broadcom committed:
Merge branch 'Code-Snipppets_D-F' into Code-Snipppets_D-F_v2

mm-broadcom committed:
Ran prettier

mm-broadcom committed:
Adjust snippet count tests

mm-broadcom committed:
Add AND and BASIS snippets

mm-broadcom committed:
Add missing BY snippets

mm-broadcom committed:
Ran Prettier, adjusted snippet counts in tests

mm-broadcom committed:
Add CORR/CORRESPONDING option

mm-broadcom commented:
@slavek-kucera @zeibura Could you both review this PR? There were a few remaining requested changes from the previous PR along with some description improvements as well.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -109,6 +115,15 @@
+  "IF x AND y AND z": {
+    "prefix": "AND",
+    "body": [
+      "IF ${1:var-1} = ${1:'AB'} AND ${3:var-2} = ${4:'CD'} THEN",
```
```suggestion
      "IF ${1:var-1} = ${2:'AB'} AND ${3:var-2} = ${4:'CD'} THEN",
```

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -45,7 +42,7 @@
   "ADD CORRESPONDING": {
     "prefix": "ADD CORRESPONDING",
-    "body": ["ADD CORR ${1:id} TO ${2:id}"],
+    "body": ["ADD ${1|CORR,CORRESPONDING|} ${2:id} TO ${3:id}"],
```
This just looks confusing to me - offering two synonyms + one option (the empty one) that does not fit the description of the snippet.
```suggestion
    "body": ["ADD CORRESPONDING ${1:id} TO ${2:id}"],
```

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -109,6 +115,15 @@
     "body": ["ALTER ${1:procName-1} TO ${2:procName-2}"],
     "description": "ALTER TO"
   },
+  "IF x AND y AND z": {
```
```suggestion
  "IF x AND y": {
```

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -136,6 +151,11 @@
+    "prefix": "BASIS",
+    "body": ["BASIS ${1:sourceName}"],
+    "description": "Similar to COPY, except that BASIS is an extended source program library statement that provides a complete COBOL program as the source for a compilation. https://www.ibm.com/docs/en/cobol-zos/6.4.0?topic=statements-basis-statement"
+  },
```
We don't support it yet. And even when we do I think we should not really advertise it.
```suggestion
```

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -188,6 +208,29 @@
+    "prefix": "BY",
+    "body": [
+      "01 ${1:TABLE-X}",
+      "\t05 ${2:RECORD-X} OCCURS ${3:100} TIMES INDEXED BY ${5:X-IND}.",
```
placeholder 4 is missing.

mm-broadcom replied:
Fixed. Thankfully it does not impact functionality for a value to be skipped, but addressed it anyways.

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -188,6 +208,29 @@
+      "\t05 ${2:RECORD-X} OCCURS ${3:100} TIMES INDEXED BY ${5:X-IND}.",
+      "\t\t10 ${6:elementary-items}"
+    ],
+    "description": "Defines an internal COBOL index for an internal COBOL table. https://publibfp.dhe.ibm.com/epubs/pdf/igy6lr40.pdf"
+  },
+  "SET x UP BY": {
+    "prefix": "BY",
+    "body": ["SET ${1:X-IND} UP BY ${2:1}"],
+    "description": "Adds +1 an internal COBOL table's index - assuming INDEXED BY. https://publibfp.dhe.ibm.com/epubs/pdf/igy6lr40.pdf"
+  },
+  "STRING x DELIMITED BY y": {
+    "prefix": "BY",
+    "body": [
+      "STRING ${1:VAR-1} DELIMITED BY ${2:VAR-2}",
+      "\tINTO ${3:VAR-3}",
+      "END-STRING."
+    ],
+    "description": "BY in the STRING (and UNSTRING) statements defines the transfer operation specifics: INTO VAR-3. https://publibfp.dhe.ibm.com/epubs/pdf/igy6lr40.pdf"
```
Isn't there a better link to provide the details than a link to the PDF documentation?

mm-broadcom replied:
@zeibura What are your thoughts on linking to the documentation online versus using the overall PDF?

zeibura replied:
I'm not entirely convinced the links are necessary at all - if they were for attribution I'd understand but apparently it's just a "for more info" link, I think we would normally assume users can find that info themselves.
If you do keep them, linking to the online doc is probably better.

slavek-kucera commented:
Please also remove the `END` snippet - it just duplicates already available entries in the completion list.

mm-broadcom committed:
Remove unneeded END snippet

END's information is already provided by the LanguageKeywords.txt from the LSP.

mm-broadcom committed:
Removed DEFAULT snippet

Removed DEFAULT snippet as QSAMBUFFINITCHAR seems to be a HLASM command rather than COBOL.

mm-broadcom committed:
Update clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Update clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Update clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Update clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>

mm-broadcom committed:
Adjust tab-stop numbers in INDEXED BY snippet

mm-broadcom committed:
Adjust snippet count tests

zeibura commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -97,7 +94,16 @@
       "END-IF."
     ],
-    "description": "IF variable-1 IS ALPHABETIC"
+    "description": "A COBOL \"class test\" used to determine whether the contents of a variable contains specific upper/lower case values. Used in conjunction with CMPR2/NOCMPR2 as compilation directives. https://www.ibm.com/docs/en/cobol-zos/6.3?topic=expressions-class-condition"
```
A COBOL \"class test\" used to determine whether the contents of a variable contain specific upper/lower case values. Used in conjunction with CMPR2/NOCMPR2 as compilation directives.

mm-broadcom replied:
That is the same description that is already there. The `\` serve as an escape character such that we can have quotes in the end result without breaking the main quotes in the JSON.

zeibura commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -97,7 +94,16 @@
+      "\t$0",
+      "END-IF."
+    ],
+    "description": "A COBOL \"class test\" which is used to determine whether the contents of a variable contains only letters and numbers (no special character) values. https://www.ibm.com/docs/en/search/alphanumeric?scope=SS6SG3_6.3.0"
```
A COBOL \"class test\" which is used to determine whether the contents of a variable contain only letters and numbers (no special characters).

zeibura commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -330,6 +368,17 @@
+      "\tCONTINUE",
+      "END-IF."
+    ],
+    "description": "The CONTINUE statement is a no-op. Indicates that no instruction is present and that execution will fall to the next sequential instruction. https://www.ibm.com/docs/en/cobol-zos/6.4?topic=statements-continue-statement"
```
The CONTINUE statement is a no-op. It indicates that no instruction is present and that execution falls to the next sequential instruction.

zeibura reviewed (commented):
see inline

mm-broadcom committed:
Description fixes

mm-broadcom committed:
Remove URLs from descriptions

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json:
```diff
@@ -18,7 +18,7 @@
     "prefix": "ACCEPT",
     "body": ["ACCEPT ${1:variable} FROM ${2|DATE,DAY,DAY-OF-WEEK,TIME|}", "$0"],
-    "description": "ACCEPT is typically used to extract the current date/time from z/OS. The other primary use case is ACCEPT from the JOB Step's SYSIN. Each line in SYSIN is tied to an individual ACCEPT statement. There are numerous options for date/time formatting: https://www.ibm.com/docs/en/cobol-zos/6.4?topic=as-date-date-yyyymmdd-day-day-yyyyddd-day-week-time / https://publibfp.dhe.ibm.com/ ...
+    "description": "ACCEPT is typically used to extract the current date/time from z/OS. The other primary use case is ACCEPT from the JOB Step's SYSIN. Each line in SYSIN is tied to an individual ACCEPT statement. There are numerous options for date/time formatting:-zos/6.4?topic=as-date-date-yyyymmdd-day-day-yyyyddd-day-week-time /"
```
There is a bunch of `-zos/6.4?topic=` and `-zos/6.3?topic=` url remnants.

mm-broadcom replied:
Has been fixed.

mm-broadcom committed:
Removed errant URL remnants

slavek-kucera commented:
One more to remove:
https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/blob/ce6b08401be13fb96c534b627fb809dacccae0f9/clients/cobol-lsp-vscode-extension/src/services/snippetcompletion/cobolSnippets.json#L1496-L1500

mm-broadcom commented:
@slavek-kucera I only removed the URLs from snippets in A-C and D-F as there are a few other branches for the rest of the alphabet to be made into PRs. It will make merging a lot easier if only the ones merged or being merged are updated.

mm-broadcom committed:
Added new snippet for DEFAULT

mm-broadcom committed:
Fix tests

mm-broadcom committed:
Removed last URL
